### PR TITLE
Port MoveIt1 PR 2682 (Fix collision normals and order of reported bodies)

### PR DESCRIPTION
**Pull request by AndyZe**
See https://github.com/ros-planning/moveit/pull/2682

This was a pretty straight-forward port

### CODE REVIEW

**Review by henningkayser — COMMENTED** [review]
Thanks for preparing this. Is there anything special to this port that requires applying it separate from the [next sync](https://github.com/ros-planning/moveit2/issues/475)?

**Comment by henningkayser** [comment]
@AndyZe could you confirm that this is correctly ported with #505 and then close this PR?

File: moveit_core/collision_detection/include/moveit/collision_detection/test_collision_common_panda.h [hunk 1]
```diff
@@ -304,7 +304,64 @@ TYPED_TEST_P(DistanceCheckPandaTest, DistanceSingle)
   }
 }
 
+template <class CollisionAllocatorType>
+class DistanceFullPandaTest : public CollisionDetectorPandaTest<CollisionAllocatorType>
```
Comment (DLu) [inline moveit_core/collision_detection/include/moveit/collision_detection/test_collision_common_panda.h:308]: In #505 this is extended from `DistanceCheckPandaTest<CollisionAllocatorType>`

File: moveit_core/collision_detection_fcl/include/moveit/collision_detection_fcl/collision_env_fcl.h [hunk 2]
```diff
@@ -154,6 +154,9 @@ class CollisionEnvFCL : public CollisionEnv
   /** \brief Callback function executed for each change to the world environment */
   void notifyObjectChange(const ObjectConstPtr& obj, World::Action action);
 
+  /** \brief Check if enable_nearest_distance_points is available */
+  void checkFCLNearestDistCapability(const DistanceRequest& req) const;
```
Comment (DLu) [inline moveit_core/collision_detection_fcl/include/moveit/collision_detection_fcl/collision_env_fcl.h:158]: This is not added to #505

File: moveit_core/collision_detection_fcl/src/collision_common.cpp [hunk 3]
```diff
@@ -585,8 +591,23 @@ bool distanceCallback(fcl::CollisionObjectd* o1, fcl::CollisionObjectd* o2, void
 #else
         dist_result.nearest_points[0] = Eigen::Vector3d(contact.pos.data.vs);
         dist_result.nearest_points[1] = Eigen::Vector3d(contact.pos.data.vs);
-        dist_result.normal = Eigen::Vector3d(contact.normal.data.vs);
+        Eigen::Vector3d normal(contact.normal.data.vs);
```
Comment (DLu) [inline moveit_core/collision_detection_fcl/src/collision_common.cpp:594]: #505 does not set the normal

**Review by DLu — COMMENTED** [review]
There are a number of places where #505 has a different implementation that you and @JafarAbdi should iron out.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 3 above)
Comment (henningkayser) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp:594]: this looks like a bug, normal is redefined in the condition below

Comment (henningkayser) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp:594]: also, the logic seems to be fixed in the sync

File: moveit_core/collision_detection_fcl/include/moveit/collision_detection_fcl/collision_env_fcl.h (hunk 2 above)
Comment (henningkayser) [reply moveit_core/collision_detection_fcl/include/moveit/collision_detection_fcl/collision_env_fcl.h:158]: this one also doesn't exist in https://github.com/ros-planning/moveit/pull/2682/files

**Comment by henningkayser** [comment]
To me it looks like the relevant changes were all added with #505 (at least with the latest update). Also, this PR does not really match the original changes. In general, backports and syncs should cherry-pick or merge the original changes to apply the same changes and maintain authorship. If there is still anything missing from #505, please add a review to that PR. Additional changes not part of the original contribution should be added in a fixup PR.

File: moveit_core/collision_detection/include/moveit/collision_detection/test_collision_common_panda.h (hunk 1 above)
Comment (henningkayser) [reply moveit_core/collision_detection/include/moveit/collision_detection/test_collision_common_panda.h:308]: https://github.com/ros-planning/moveit/pull/2682/files also does so

**Comment by AndyZe** [comment]
If I'd known that MoveIt1 PR's were still being systematically ported to MoveIt2, I wouldn't have opened this PR.

There are some small things I like better about this version. I think the function name `checkFCLNearestDistCapability()` is better than `checkFCLCapabilites()`, for example. But, I don't care enough to actually finish the PR   ;)

**Pull request closed without merging**